src/literal.h

```diff
@@ -158,24 +191,199 @@ class Literal {
+  Literal shrUI32x4(const Literal& other) const;
+  Literal addI32x4(const Literal& other) const;
+  Literal subI32x4(const Literal& other) const;
+  Literal mulI32x4(const Literal& other) const;
+  Literal negI64x2() const;
+  Literal anyTrueI64x2() const;
+  Literal allTrueI64x2() const;
+  Literal shlI64x2(const Literal& other) const;
+  Literal shrSI64x2(const Literal& other) const;
+  Literal shrUI64x2(const Literal& other) const;
+  Literal addI64x2(const Literal& other) const;
+  Literal subI64x2(const Literal& other) const;
+  Literal absF32x4() const;
+  Literal negF32x4() const;
+  Literal sqrtF32x4() const;
+  Literal addF32x4(const Literal& other) const;
+  Literal subF32x4(const Literal& other) const;
+  Literal mulF32x4(const Literal& other) const;
+  Literal divF32x4(const Literal& other) const;
+  Literal minF32x4(const Literal& other) const;
+  Literal maxF32x4(const Literal& other) const;
+  Literal absF64x2() const;
+  Literal negF64x2() const;
+  Literal sqrtF64x2() const;
+  Literal addF64x2(const Literal& other) const;
+  Literal subF64x2(const Literal& other) const;
+  Literal mulF64x2(const Literal& other) const;
+  Literal divF64x2(const Literal& other) const;
+  Literal minF64x2(const Literal& other) const;
+  Literal maxF64x2(const Literal& other) const;
+  Literal truncSatToSI32x4() const;
+  Literal truncSatToUI32x4() const;
+  Literal truncSatToSI64x2() const;
+  Literal truncSatToUI64x2() const;
+  Literal convertSToF32x4() const;
+  Literal convertUToF32x4() const;
+  Literal convertSToF64x2() const;
+  Literal convertUToF64x2() const;
+
+ private:
+  Literal addSatSI8(const Literal& other) const;
+  Literal addSatUI8(const Literal& other) const;
+  Literal addSatSI16(const Literal& other) const;
+  Literal addSatUI16(const Literal& other) const;
+  Literal subSatSI8(const Literal& other) const;
+  Literal subSatUI8(const Literal& other) const;
+  Literal subSatSI16(const Literal& other) const;
+  Literal subSatUI16(const Literal& other) const;
 };
 
 } // namespace wasm
 
 namespace std {
 template<> struct hash<wasm::Literal> {
   size_t operator()(const wasm::Literal& a) const {
+    uint8_t bytes[16] = {};
+    a.getBits(bytes);
+    int64_t chunks[2];
+    memcpy(chunks, bytes, sizeof(chunks));
     return wasm::rehash(
-      uint64_t(hash<size_t>()(size_t(a.type))),
-      uint64_t(hash<int64_t>()(a.getBits()))
+      wasm::rehash(
+        uint64_t(hash<size_t>()(size_t(a.type))),
+        uint64_t(hash<int64_t>()(chunks[0]))
+      ),
+      uint64_t(hash<int64_t>()(chunks[1]))
     );
   }
 };
 template<> struct less<wasm::Literal> {
   bool operator()(const wasm::Literal& a, const wasm::Literal& b) const {
     if (a.type < b.type) return true;
     if (a.type > b.type) return false;
-    return a.getBits() < b.getBits();
+    switch (a.type) {
+      case wasm::Type::i32: return a.geti32() < b.geti32();
+      case wasm::Type::f32: return a.getf32() < b.getf32();
```

Comment:
I used `getBits` (with the void* change) here since the literals have to be reinterpreted as the correct type before their bits can be extracted via `geti32` and friends. Happy to change this if we decide to go the other direction with `getBIts`.